File: 2559.count-vowel-strings-in-ranges/solution.kt
class Solution {
    fun vowelStrings(words: Array<String>, queries: Array<IntArray>): IntArray {
        val prefixSum = words.runningFold(0) { acc, word -> 
            if (word.isNotEmpty() && word.first() in "aeiou" && word.last() in "aeiou") {
                acc + 1
            } else {
                acc
            }
        }
        val results = queries.map { (l, r) -> prefixSum[r + 1] - prefixSum[l] }
        return results.toIntArray()
    }
}

File: 2559.count-vowel-strings-in-ranges/solution.go
func vowelStrings(words []string, queries [][]int) []int {
    res := make([]int, 0, len(queries))
    pre := prefixSum(words)
    for _, q := range queries {
        res = append(res, pre[q[1] + 1] - pre[q[0]])
    }
    return res
}

func prefixSum(words []string) []int {
    res := make([]int, 1, len(words) + 1)
    count := 0
    for _, w := range words {
        if len(w) > 0 && isVowel(w[0]) && isVowel(w[len(w) - 1]) {
            count++
        }
        res = append(res, count)
    }
    return res
}

func isVowel(ch byte) bool {
    return strings.IndexByte("aeiou", ch) != -1
}